### PR TITLE
Orchestrator save/delete readme file to flat container

### PR DESCRIPTION
**Pull request by lyndaidaii**
* PackageStatusProcessor added logic to save readme files to flat container
* Added telemetry for readme file operations.
* ServerCommon version upgrade. (This need to be updated after https://github.com/NuGet/NuGetGallery/pull/8377 is merged)

### CODE REVIEW

File: src/NuGet.Jobs.GitHubIndexer/NuGet.Jobs.GitHubIndexer.csproj [hunk 1]
```diff
@@ -93,7 +93,7 @@
       <Version>0.32.0</Version>
     </PackageReference>
     <PackageReference Include="NuGetGallery.Core">
-      <Version>4.4.5-dev-4220086</Version>
+      <Version>4.4.6-gallery-2configuration-4385211</Version>
```
Comment (lyndaidaii) [inline src/NuGet.Jobs.GitHubIndexer/NuGet.Jobs.GitHubIndexer.csproj:96]: Need to be updated after https://github.com/NuGet/NuGetGallery/pull/8377 is merged

File: src/NuGet.Services.Validation.Orchestrator/PackageStatusProcessor.cs [hunk 2]
```diff
@@ -43,13 +46,27 @@ public class PackageStatusProcessor : EntityStatusProcessor<Package>
...
+                }
+            }
         }
 
         protected override async Task OnCleanupAfterDatabaseUpdateFailure(
             IValidatingEntity<Package> validatingEntity,
             PackageValidationSet validationSet)
         {
-            if (validatingEntity.EntityRecord.EmbeddedLicenseType != EmbeddedLicenseFileType.Absent)
+            if (validatingEntity.EntityRecord.EmbeddedLicenseType != EmbeddedLicenseFileType.Absent || validatingEntity.EntityRecord.HasEmbeddedReadme)
```
Comment (lyndaidaii) [inline src/NuGet.Services.Validation.Orchestrator/PackageStatusProcessor.cs:69]: Remove condition `validatingEntity.EntityRecord.HasEmbeddedReadme`

File: src/NuGet.Services.Validation.Orchestrator/PackageStatusProcessor.cs [hunk 3]
```diff
@@ -43,13 +46,27 @@ public class PackageStatusProcessor : EntityStatusProcessor<Package>
                     _logger.LogInformation("Successfully extracted the license file.");
                 }
             }
+
+            if (validatingEntity.EntityRecord.HasEmbeddedReadme)
+            {
+                using (_telemetryService.TrackDurationToExtractReadmeFile(validationSet.PackageId, validationSet.PackageNormalizedVersion, validationSet.ValidationTrackingId.ToString()))
+                using (var packageStream = await _packageFileService.DownloadPackageFileToDiskAsync(validationSet))
```
Comment (joelverhagen) [inline src/NuGet.Services.Validation.Orchestrator/PackageStatusProcessor.cs:53]: We are downloading the package above. We shouldn't need to download it again. Consider putting both "extract" steps in the same downloading `using`.

File: src/NuGet.Services.Validation.Orchestrator/PackageStatusProcessor.cs [hunk 4]
```diff
@@ -43,13 +46,27 @@ public class PackageStatusProcessor : EntityStatusProcessor<Package>
...
+                {
+                    _logger.LogInformation("Extracting the readme file of type {EmbeddedReadmeType} for the package {PackageId} {PackageVersion}",
+                        validatingEntity.EntityRecord.EmbeddedReadmeType,
+                        validationSet.PackageId,
+                        validationSet.PackageNormalizedVersion);
+                    await _coreReadmeFileService.ExtractAndSaveReadmeFileAsync(validatingEntity.EntityRecord, packageStream);
```
Comment (joelverhagen) [inline src/NuGet.Services.Validation.Orchestrator/PackageStatusProcessor.cs:59]: just a note: this only saves to global storage, not China storage so if we ever want to make this V3 asset public/documented we would need to move this logic to catalog2dnx

File: src/NuGet.Services.Validation.Orchestrator/PackageStatusProcessor.cs (hunk 3 above)
Comment (skofman1) [reply src/NuGet.Services.Validation.Orchestrator/PackageStatusProcessor.cs:53]: +1

File: src/NuGet.Services.Validation.Orchestrator/PackageStatusProcessor.cs (hunk 4 above)
Comment (skofman1) [reply src/NuGet.Services.Validation.Orchestrator/PackageStatusProcessor.cs:59]: Yup, but I don't think having it in China storage is a requirement right now.

Comment (joelverhagen) [inline src/NuGet.Services.Validation.Orchestrator/PackageStatusProcessor.cs:59]: Is the catalog2dnx work planned for soon? This code shouldn't go live until catalog2dnx is handling clean-up for when a package is deleted.

Package deletes do not flow into orchestrator at all.

**Review by joelverhagen — CHANGES_REQUESTED** [review]
⌚

**Comment by zhhyu** [comment]
I am not sure why we change the decision to use Orchestrator rather than Catalog2Dnx here. I have same questions as above:
1. We still need the Catalog2Dnx job to clean up if the package should be deleted. If we put all the work (placing the readme into the flat container and deleting it from the flat container) there, it will be easier to maintain. I think the reason to place the license file using Orchestrator and show it to customers asap is because of some legal concerns;
2. One reason that we'd like to use the flat container is that we have a separate storage in China, and PM has some thoughts to support showing readmes in IDEs one day.
3. We are refreshing the Orchestrator. So maybe keeping this change inside the V3 pipeline will be easier.

File: src/NuGet.Services.Validation.Orchestrator/Job.cs [hunk 5]
```diff
@@ -474,6 +475,11 @@ private static void ConfigureFlatContainer(ContainerBuilder builder)
                 .RegisterType<CoreLicenseFileService>()
                 .WithKeyedParameter(typeof(ICoreFileStorageService), CoreLicenseFileServiceBindingKey)
                 .As<ICoreLicenseFileService>();
+
+            builder
+                .RegisterType<CoreReadmeFileService>()
+                .WithKeyedParameter(typeof(ICoreFileStorageService), CoreReadmeFileServiceBindingKey)
```
Comment (loic-sharma) [inline src/NuGet.Services.Validation.Orchestrator/Job.cs:481]: Is this right? It looks like you want the `ICoreFileStorage` that reads from the flat container, which currently uses the `CoreLicenseFileServiceBindingKey` key:

https://github.com/NuGet/NuGet.Jobs/blob/7dbf54e7a4c5d87a287ee1c171b74e7218f5ef47/src/NuGet.Services.Validation.Orchestrator/Job.cs#L455-L468

**Review by joelverhagen — APPROVED** [review]
Use official build of NuGetGallery.Core then :shipit:

File: src/NuGet.Services.Validation.Orchestrator/Telemetry/TelemetryService.cs [hunk 6]
```diff
@@ -35,8 +35,9 @@ public class TelemetryService : ITelemetryService, ISubscriptionProcessorTelemet
         private const string MessageHandlerDurationSeconds = OrchestratorPrefix + "MessageHandlerDurationSeconds";
         private const string MessageLockLost = OrchestratorPrefix + "MessageLockLost";
         private const string SymbolsMessageEnqueued = OrchestratorPrefix + "SymbolsMessageEnqueued";
-        private const string ExtractLicenseFileDuration = OrchestratorPrefix + "ExtractLicenseFileDuration";
+        private const string ExtractLicenseAndReadmeFileDuration = OrchestratorPrefix + "ExtractLicenseAndReadmeFileDuration";
```
Comment (loic-sharma) [inline src/NuGet.Services.Validation.Orchestrator/Telemetry/TelemetryService.cs:38]: Could you add a suffix to clarify the unit here?

```suggestion
        private const string ExtractLicenseAndReadmeFileDurationSeconds = OrchestratorPrefix + "ExtractLicenseAndReadmeFileDurationSeconds";
```

**Review by loic-sharma — APPROVED** [review]

**Comment by loic-sharma** [comment]
Nice work! :)

**Pull request merged**